### PR TITLE
Integrate triton moe kernel

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

This PR is to follow up https://github.com/sgl-project/sglang/issues/7287.
The main purpose is to replace fused_moe with Triton v3.4.0 matmul_ogs kernel, in order to improve fused_moe's performance.

```
#python ./benchmark/kernels/fused_moe_triton/benchmark_sglang_fused_moe_triton.py --use-cuda-graph
INFO 07-01 07:27:05 [__init__.py:244] Automatically detected platform cuda.
shape_configs={'num_experts': 128, 'topk': 8, 'hidden_size': 2048, 'shard_intermediate_size': 768, 'dtype': torch.bfloat16, 'block_shape': None}
benchmark sglang_fused_moe_triton_v340 with batch_size=128
benchmark sglang_fused_moe_triton with batch_size=128
Using default MoE kernel config. Performance might be sub-optimal! Config file not found at /usr/local/lib/python3.10/dist-packages/sglang/srt/layers/moe/fused_moe_triton/configs/triton_3_4_0/E=128,N=384,device_name=NVIDIA_L20Y.json, you can create them with https://github.com/sgl-project/sglang/tree/main/benchmark/kernels/fused_moe_triton
benchmark sglang_fused_moe_triton_v340 with batch_size=256
benchmark sglang_fused_moe_triton with batch_size=256
benchmark sglang_fused_moe_triton_v340 with batch_size=512
benchmark sglang_fused_moe_triton with batch_size=512
benchmark sglang_fused_moe_triton_v340 with batch_size=1024
benchmark sglang_fused_moe_triton with batch_size=1024
benchmark sglang_fused_moe_triton_v340 with batch_size=2048
benchmark sglang_fused_moe_triton with batch_size=2048
benchmark sglang_fused_moe_triton_v340 with batch_size=4096
benchmark sglang_fused_moe_triton with batch_size=4096
benchmark sglang_fused_moe_triton_v340 with batch_size=8192
benchmark sglang_fused_moe_triton with batch_size=8192
fused-moe-performance:
   batch_size  sglang_fused_moe_triton_v340  sglang_fused_moe_triton
0       128.0                      0.259136                 0.241552
1       256.0                      0.255424                 0.303200
2       512.0                      0.269712                 0.342176
3      1024.0                      0.303376                 0.395168
4      2048.0                      0.366592                 0.516064
5      4096.0                      0.530272                 0.881728
6      8192.0                      0.891840                 1.615424
```
image
Unit test precision check passed:
```
[root@b75ba454e0f7 /sgl-workspace/sglang]
#python ./test/srt/test_triton_fused_moe.py
INFO 07-02 09:14:45 [__init__.py:244] Automatically detected platform cuda.
Running MoE tests: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 48/48 [00:01<00:00, 29.36it/s]
.
----------------------------------------------------------------------
Ran 1 test in 1.636s

OK

```

## Modifications

The w13 and w2's shape required by matmul_ogs is different with the shape of current Triton fused_moe_kernel. It needs to do transpose in load weight phase before GEMM. 

Added benchmark test with cuda graph capture enable and precision test case.

Note: This PR requires Triton v3.4.0 and triton_kernels installed. Moreover, it needs an additional hot-fix in Triton v3.4.0. I'll  issue a separate PR of this fix to Triton-lang repo later.

Triton hot-fix PR: https://github.com/triton-lang/triton/pull/7367

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    w1_scale: Optional[torch.Tensor] = None,
+    w2_scale: Optional[torch.Tensor] = None,
+    a1_scale: Optional[torch.Tensor] = None,
+    a2_scale: Optional[torch.Tensor] = None,
+    block_shape: Optional[list[int]] = None,
+) -> torch.Tensor:
+    return torch.empty_like(hidden_states)
+
+
+"""
```
Comment (ispobock) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: remove this comment

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    assert hidden_states.ndim == 2, "hidden_states must be 2D"
+    assert (
+        hidden_states.shape[-1] == w1.shape[-2]
+    ), f"hidden_states shape[-1] {hidden_states.shape} must be equal to w1 shape[-2] {w1.shape}"
+    assert (
+        w2.shape[-1] == w1.shape[1]
+    ), f"w2 shape[-1] {w2.shape[-1]} must be equal to w1 shape[1] {w1.shape[1]}"
+
+    # feature check
+    # assert inplace == False, "Inplace is not supported in new triton MoE kernel"
```
Comment (ispobock) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: why comment this assert?

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    global_num_experts: int = -1,
+    expert_map: Optional[torch.Tensor] = None,
+    w1_scale: Optional[torch.Tensor] = None,
+    w2_scale: Optional[torch.Tensor] = None,
+    a1_scale: Optional[torch.Tensor] = None,
+    a2_scale: Optional[torch.Tensor] = None,
+    block_shape: Optional[list[int]] = None,
+) -> torch.Tensor:
+
+    if not renormalize:
```
Comment (ispobock) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:33]: How to handle renormalize case?

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+import sys
+from typing import Callable, List, Optional, Tuple
+
+import torch
+from triton_kernels.matmul_ogs import matmul_ogs
+from triton_kernels.routing import GatherIndx, RoutingData, ScatterIndx, routing
+
+# from vllm.utils import direct_register_custom_op
```
Comment (ispobock) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: remove this line

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (ispobock) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: Add source reference

File: python/sglang/srt/server_args.py [hunk 6]
```diff
@@ -1517,6 +1518,11 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Enable returning hidden states with responses.",
         )
+        parser.add_argument(
```
Comment (ispobock) [inline python/sglang/srt/server_args.py:1558]: --enable-triton-kernel-moe?

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    apply_router_weight_on_input: bool = False,
+    use_fp8_w8a8: bool = False,
+    per_channel_quant: bool = False,
+    global_num_experts: int = -1,
+    expert_map: Optional[torch.Tensor] = None,
+    w1_scale: Optional[torch.Tensor] = None,
+    w2_scale: Optional[torch.Tensor] = None,
+    a1_scale: Optional[torch.Tensor] = None,
+    a2_scale: Optional[torch.Tensor] = None,
+    block_shape: Optional[list[int]] = None,
```
Comment (hlu1) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:30]: Please add assertions for unused input args

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        gather_indx=gather_indx,
+        gammas=routing_data.gate_scal if apply_router_weight_on_input else None,
+    )
+
+    from sgl_kernel import gelu_and_mul, silu_and_mul
+
+    if activation == "silu":
+        silu_and_mul(intermediate_cache1.view(-1, N), intermediate_cache2)
+    elif activation == "gelu":
+        gelu_and_mul(intermediate_cache1.view(-1, N), intermediate_cache2)
```
Comment (hlu1) [inline python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:132]: Are you using the sglang activations kernels because of the weight shuffling required by the openai kernels?

**Comment by xutizhou** [comment]
Hi @yuan-luo , when you used Triton v3.4.0, did you encounter an import error when launching the server? Since SGLang supports Triton v3.3, do you have any insights on how to fix this issue?
image

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 1 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: Removed.

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 3 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:33]: If renormalize=True, the host side will not do softmax, it will leave the Triton kernel inside to to softmax.
```
    if not renormalize:
        gating_output = torch.softmax(gating_output, dim=-1)
    routing_data, gather_idx, scatter_idx = routing(gating_output, topk, renormalize)
```

In triton kernel routing(), if renormalize=True, namely sm_first=True, it will do softmax inside.
If renormalize=False, sm_first is false, it will skip softmax.
```
def routing(logits, n_expts_act, sm_first=False, expt_indx=None, simulated_ep=1, n_rows=None):
    from .topk import topk
    if sm_first:
        logits = torch.softmax(logits, dim=-1)
    expt_scal, expt_indx, bitmatrix = topk(logits, n_expts_act,  #
                                           apply_softmax=not sm_first, y_indx=expt_indx, n_rows=n_rows)
    # mutate bitmatrix
    if simulated_ep > 1:
        expt_scal, expt_indx, bitmatrix = prune_routing(expt_scal, expt_indx, bitmatrix, simulated_ep)
    hist, topk_indx, gate_indx, gate_scal = sort_tokens(expt_scal, expt_indx, bitmatrix)
    # pack the matmul data structure
    n_expts_tot = logits.shape[-1] // simulated_ep
    gather_indx = GatherIndx(src_indx=topk_indx, dst_indx=gate_indx)
    scatter_indx = ScatterIndx(src_indx=gate_indx, dst_indx=topk_indx)
    expt_data = compute_expt_data(hist, n_expts_tot, topk_indx.numel())
    return RoutingData(gate_scal, hist, n_expts_tot, n_expts_act, expt_data), gather_indx, scatter_indx
```

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 4 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: Done.

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 8 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:132]: Here need to invoke activation kernels, I just pick sglang version.

**Comment by yuan-luo** [comment]
> Hi @yuan-luo , when you used Triton v3.4.0, did you encounter an import error when launching the server? Since SGLang supports Triton v3.3, do you have any insights on how to fix this issue? image

Aha, @xutizhou  I encountered this issue and I forgot to mention it. Just modify this triton file: triton/runtime/cache.py
Add the following part in the head.
```
from .. import knobs   <<<<<<<<< After this line


from pathlib import Path

def get_home_dir():
    return os.getenv("TRITON_HOME", Path.home())


def default_cache_dir():
    return os.path.join(get_home_dir(), ".triton", "cache")


def default_override_dir():
    return os.path.join(get_home_dir(), ".triton", "override")


def default_dump_dir():
    return os.path.join(get_home_dir(), ".triton", "dump")


class CacheManager(ABC):   <<<<<<<<<<<< Before this line

    def __init__(self, key, override=False, dump=False):
        pass
```

I'll issue another PR to triton-lang later.

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 2 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py]: Undo the comment.

**Comment by xutizhou** [comment]
> triton-lang/triton#7367

hi @yuan-luo, thank you so much for prompt reply. I will give it a try.

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 7 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:30]: Added.

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    use_fp8_w8a8: bool = False,
+    per_channel_quant: bool = False,
+    global_num_experts: int = -1,
+    expert_map: Optional[torch.Tensor] = None,
+    w1_scale: Optional[torch.Tensor] = None,
+    w2_scale: Optional[torch.Tensor] = None,
+    a1_scale: Optional[torch.Tensor] = None,
+    a2_scale: Optional[torch.Tensor] = None,
+    block_shape: Optional[list[int]] = None,
+) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:31]: Added assertion.

File: python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py (hunk 3 above)
Comment (ispobock) [reply python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py:33]: I see.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 10]
```diff
@@ -438,9 +465,11 @@ def __init__(
         self.inplace = inplace
         self.no_combine = no_combine
 
+        self.use_triton_kernels = global_server_args_dict["enable_triton_kernel_moe"]
```
Comment (Alcanderian) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: CPU do not support triton

Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Fixed.

**Review by ispobock — APPROVED** [review]

**Comment by ispobock** [comment]
@yuan-luo please fix the lint

**Comment by yuan-luo** [comment]
> @yuan-luo please fix the lint

Done.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 11]
```diff
@@ -97,22 +110,23 @@ def create_weights(
...
-                num_experts, 2 * intermediate_size, hidden_size, dtype=params_dtype
-            ),
-            requires_grad=False,
+        w13_weight_raw = torch.empty(
+            num_experts, 2 * intermediate_size, hidden_size, dtype=params_dtype
         )
+        if self.use_triton_kernels:
+            w13_weight_raw = w13_weight_raw.transpose(-2, -1).contiguous()
+        w13_weight = torch.nn.Parameter(w13_weight_raw, requires_grad=False)
+
```
Comment (Alcanderian) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]:
```suggestion
        w13_weight_n, w13_weight_k = 2 * intermediate_size, hidden_size
        if self.use_triton_kernels:
            w13_weight_n, w13_weight_k = w13_weight_k, w13_weight_n
        w13_weight = torch.nn.Parameter(
            torch.empty(
                num_experts, w13_weight_n, w13_weight_k, dtype=params_dtype
            ),
            requires_grad=False,
        )
```

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 12]
```diff
@@ -630,6 +663,8 @@ def _load_w2(
             )
         else:
             if not self.use_presharded_weights:
+                if self.use_triton_kernels:
+                    loaded_weight = loaded_weight.transpose(-2, -1).contiguous()
```
Comment (Alcanderian) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]:
```suggestion
                    loaded_weight = loaded_weight.transpose(-2, -1)
```

Comment (Alcanderian) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Using contiguous will trigger tensor allocation, better to avoid this.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 11 above)
Comment (Alcanderian) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Using contiguous will trigger tensor allocation, better to avoid this.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 12 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Make sense. Fixed.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 11 above)
Comment (yuan-luo) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Make sense. Fixed.

**Review by Alcanderian — APPROVED** [review]

**Review by merrymercy — APPROVED** [review]

**Comment by zhyncs** [comment]
Let's wait for the CIs :)

**Comment by dongyibo** [comment]
I found that to use this kernel, the value of topk must be a power of 2. Is that correct? @yuan-luo
